### DISCUSSION TITLE
Query Database from Form dialog

### DISCUSSION THREAD

**Discussion by rhemasystems** [serenity-is/Serenity#6056] [Help] [answered]
Please I have a form that receivs user input, one of the fields in a free text fieled but how can i validate what has been entered  against the database when the user clicks the Save button before actually saving in the database.

![image](https://user-images.githubusercontent.com/50826372/138707986-4b3c4941-9385-40d7-ab8d-75b035c70780.png)

**Comment by VictorTomaili** [comment, marked as answer]
Don't do this on the client side. Anyone who knows the `bypassing validation rule` can create duplicate records that can break your data consistency. 
The best practice is to make serial numbers unique in the database.

**Reply by rhemasystems** [reply]
Thank you Victor, the Serialnumber is unique in the table, the logic is sending serialnumber  from one table to another table, but the user has to enter first the serialnumber, gets validated in the original table if the serialnumber entered is correct before saving in the other table

**Reply by VictorTomaili** [reply]
use lookup editor for serial numbers, if you have thousands of serial numbers or more, you can use servicelookupeditor

**Reply by rhemasystems** [reply]
I have found a way out, my challenge now is how to filter based on the product name selected on the same form because there is a productid foreign key in the serial table,so the lookup should filter only serials of the selected product

**Reply by rhemasystems** [reply]
I finally got it to work, thanks Victor

**Reply by rhemasystems** [reply]
Now another issue, the serialnumber after using the servicelookup to select it doesnt save in the table.

**Comment by luongnv6** [comment]
Check in validaterequest in Respository
```C#
private class MySaveHandler : SaveRequestHandler<MyRow> {
            protected override void ValidateRequest()
            {
                base.ValidateRequest();
                if (IsCreate || (IsUpdate && (Row.SerialNumber!= Old.SerialNumber)))
                    if (!Connection.Exists<SerialNumbersRow>(fld.SerialNumber== Row.SerialNumber.Trim()))
                        throw new ValidationError("Not Already Exist", "SerialNumber", $"Serial Number\"{Row.SerialNumber}\" not exists.");
            }
        }
`

**Reply by rhemasystems** [reply]
Thank you luongnv6 , it worked perfectly, but my question is, cant I achieve this with validateEntity method in the XYZEditor.ts because I have to do this for every row in the grid

**Reply by luongnv6** [reply]
write in contructor:
```C#
this.form.SerialNumber.change(e=>{
        if(!Q.isEmptyOrNull(this.form.SerialNumber.value)){
               var item = Q.getLookup<SerialNumberRow>(lookupKey of SerialNumber).itemById[this.form.SerialNumber.value];
              if(item == undefined)
                     Q.alert("Serial Number is wrong");
       }
});
`

**Reply by luongnv6** [reply]
or
```C#
this.form.SerialNumber.addValidationRule(this.uniqueName, e => {
       if(!Q.isEmptyOrNull(this.form.SerialNumber.value)){
                var item = Q.getLookup(lookupKey of SerialNumber).itemById[this.form.SerialNumber.value];
                 if(item == undefined)
                      return "Serial Number is wrong";
       }
 });
`

**Reply by luongnv6** [reply]
or
```C#
this.form.SerialNumber.addValidationRule(this.uniqueName, e => {
       if(!Q.isEmptyOrNull(this.form.SerialNumber.value)){
                SerialNumberService.List({ ContainsField: "SerialNumber", ContainsText: this.form.SerialNumber.value}, res=>{
                      if(res.TotalCount == 0)
                      return "Serial Number is wrong";                 
               })
       }
 });

**Reply by rhemasystems** [reply]
Thank you so much, the last option with the Service worked for me

> 
> 
> or
> 
> ```cs
> this.form.SerialNumber.addValidationRule(this.uniqueName, e => {
>        if(!Q.isEmptyOrNull(this.form.SerialNumber.value)){
>                 SerialNumberService.List({ ContainsField: "SerialNumber", ContainsText: this.form.SerialNumber.value}, res=>{
>                       if(res.TotalCount == 0)
>                       return "Serial Number is wrong";                 
>                })
>        }
>  });
> ```

Thank you so much, this option worked for me
